### PR TITLE
PIN 14: Event-driven Flows

### PR DESCRIPTION
**Pull request by cicdw**
This PIN supersedes PIN 8 with a new proposed implementation for event driven flows that reuses many of the new hooks that have been introduced since PIN 8 was first proposed, including Clocks and idempotency keys.

Any and all suggestions / edits welcome!

### CODE REVIEW

**Comment by jlowin** [comment]
👏👏 Excellent.

A few thoughts to shore up:
- the proposed `EventParameter` feels like it could formalize a process ahead of solid use cases. Perhaps we could start by placing the event in context, which would allow us to trivially add a task in the future that retrieved and returned it for first-class dataflow. (I'd also like to do this for the event that kicks off regularly-scheduled flows!)
- Could you elaborate on why the `next()` method is unimplemented? Could it just block until the next event?

**Comment by cicdw** [comment]
@jlowin so you're suggesting we just put the event into context for any task to use, instead of elevating it to a Parameter?  That certainly works, it just _feels_ like an abuse of context (e.g., we'll need to confirm that things like retries work, although I think they will since context is attached to the flow run).  That being said, it would be a very simple implementation so I'm not opposed to it.

For `next()`, I have two answers:
- in order for event clocks to work seamlessly with time-based clocks, `next` can't block (but it can do a poll)
- on the cloud-side, we can't block on events because we don't have access to the logic for polling for events; consequently, with the above point, returning nothing felt like the appropriate implementation

File: docs/core/PINs/PIN-14-Listener-Flows-2.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+Subclasses of this `EventClock` can then be used to construct an appropriate `Schedule` for the Flow (possibly with other time-driven clocks!), and `flow.run` will essentially work with minimal change.  The only additional piece of information that needs to be conveyed is the value for the designated event parameter, and this can be achieved through one additional call (maybe `flow.retrieve_event()` which then extracts the event from the `EventClock`) and then placing the result into Prefect context.
+
+In addition, a new `listen` method on `FlowRunner` classes will be responsible for calling the schedule and configuring context for each run.  Much of the current scheduling logic in `flow.run()` should be replaceable by this new method on the `FlowRunner` class.
+
+### Cloud
+
+There are a few proposed changes that will also require changes to Prefect Cloud:
+
+- any subclass of the `EventClock` will be deserialized as the base `EventClock` class, which returns an empty list for `next`
+- users will be responsible for starting and stopping the listening by creating / cancelling a given Flow run
```
[the remark this answers is not in the document]
Comment (cicdw) [reply docs/core/PINs/PIN-14-Listener-Flows-2.md]: `Listening` will be analogous (and treated essentially identically to) `Running`, so things like the Lazarus process will always make sure the Flow is in an active listening state.

File: docs/core/PINs/PIN-14-Listener-Flows-2.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+- once created, an appropriate agent will pick up the flow run and submit it for "execution" (in this case, submit it to listen); we might need to introduce a new pre-listening state so that the Agent knows this is a listener and not a normal flow run
+- During this execution, the Flow run will enter a `Listening` state and `CloudFlowRunner.listen()` will be used to poll for events; anytime an event is found, it will be used to call the Cloud API and create a new parametrized flow run with the appropriate event parameter payload
```
[the remark this answers is not in the document]
Comment (cicdw) [reply docs/core/PINs/PIN-14-Listener-Flows-2.md]: Yea, agreed

**Comment by cicdw** [comment]
@jlowin any additional thoughts here?

**Pull request merged**